junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodBasedTestDescriptor.java

```diff
@@ -81,7 +87,24 @@ public final Set<TestTag> getTags() {
 
 	@Override
 	public Set<ExclusiveResource> getExclusiveResources() {
-		return getExclusiveResourcesFromAnnotation(getTestMethod());
+		// @formatter:off
+		Function<ResourceLocksProvider, Set<ResourceLocksProvider.Lock>> providerToLocks =
+				p -> p.provideForMethod(getTestClass(), getTestMethod());
+
+		Set<ExclusiveResource> fromMethodLevelAnnotation = getExclusiveResourcesFromAnnotation(
+				getTestMethod(),
+				providerToLocks
+		);
+		Stream<ExclusiveResource> fromProvidersInClassLevelAnnotation = getExclusiveResourcesFromAnnotationProviders(
+				getTestClass(),
```

Comment:
What about `@Nested` test classes? I think we'll have to collect exclusive resources from providers from the test class and, if it's a `@Nested` one, also from its enclosing class, recursively.

Reply:
On second thought, there is no need to apply locks from the class level to methods as it is redundant. When executing the method, the class-level locks are already held.

Reply:
> What about `@Nested` test classes? I think we'll have to collect exclusive resources from providers from the test class and, if it's a `@Nested` one, also from its enclosing class, recursively.

I suppose that the behaviour of `ResourceLocksProvider` should be similar to the plain `@ResourceLock("...")`.
Let's consider such an example:
```java
@ResourceLock("A")
static class TestClass {

    @Nested
    @ResourceLock("B")
    class NestedTestClass {
    }
}
```
`NestedTestClass` gets only `B` but **does not** get `A`.

That's why I'm not sure that `@Nested` classes should invoke providers of enclosing classes.

Reply:
Please disregard my initial comment. We should neither check the test class nor any of its enclosing classes since that's already being done by the parent test descriptors. Thus, please remove the extra code that sets `fromProvidersInClassLevelAnnotation` etc.

Reply:
If we remove this code then `LocksProvider#provideForMethod()` will not be invoked for `test()`:
```java
@ResourceLock(providers = LocksProvider.class)
static class TestClass {

    @Test
    void test() {
    }
}
```

Reply:
Hmm, I see. I didn't think about that use case. In that case, we also need to ensure `LocksProvider#provideForMethod()` gets called for nested classes, e.g.:

```java
@ResourceLock(providers = LocksProvider.class)
static class TestClass { // A: `LocksProvider#provideForClass()` is called
	@Nested
	class Inner { // B: `LocksProvider#provideForNestedClass()` is called
    	@Test
    	void test() { // C: `LocksProvider#provideForMethod()` is called
    	}
	}
}
```

If I'm reading the current code correctly, neither B nor C would be called.

Reply:
You're right.
Done